### PR TITLE
Atom/mriegger/fullscreenshadow2

### PR DESCRIPTION
**Pull request by mrieggeramzn**
Full screen shadows. This aims to improve performance by extracting the shadow code into a fullscreen pass. This pass outputs a R16_FLOAT buffer than contains boolean Is-In_Shadow test result that the forward pass can read from.

Savings should be due to lesser VGPR usage in the forward shader along with shadow coords no longer interpolated from the vertex shaders.

Loft perf improvements:
This adds 0.485 MS for the fullscreen shadow calculation.
Loft forward pass improved from 14.90 ms to a theoretical best of 9.5 ms in a earlier Pix capture. Note I said theoretical because I had to hard code a branch to get around the fact that we don't have shader-variants enabled in Loft. Without my branch hack the performance was still ~12 ms so still a net win

(Card was 2080, I was profiling Loft via Samples.GameLauncher.exe with run parameters
--project-path=D:\fullscreenshadowlatest2\o3de\o3de-samples-project\Samples  +loadlevel levels/archvis/interior_03/interior_03.spawnable --enablePixGpuMarkers)

UPDATE from antonmic
Made multiple changes and fixes to the fullscreenshadow technique, quality is now in a pretty good state with default settings, and options like PCF can be easily selected for different quality tradeoffs. List of changes:
-Fixed world normal calculation in fullscreen shadow shader (was not working at all previously)
-Further improved normal calculation by checking right and left, top and bottom and picking the smallest depth gradients for the normal calculation. This is a form of edge rejection that is also used with our SSAO.
-Disabled slope bias calculation as it has visible artifacts both with and without the fullscreen shadow technique. Instead bumped the default normal bias value to 2.5 which gets rid of a most minor artifacts while still being conservative
-Introduce a fullscreen depth aware blur on the fullscreen shadow buffer. Noticeably smooths out shadowmap blockiness without any of the drawbacks of PCF or ESM. However if the camera gets too close an objects the blockiness can be spotted despite the blur. Some hybrid where close to camera pixels use PCF can be investigated in the future.
-Introduced new debug variables in the debug render component that are accessible in shaders. These are general purpose and provide shader authors a means of immediately getting bools and floats from the editor to a shader for debug purposes without having to do any plumbing. Used this extensively whilst debugging shadows.
-Introduce new debug view mode in the debug render component for visualizing cascaded shadows.

Since all previous commits were reviewed by myself and invertednormal, please review the changes I made with the following commit [here](https://github.com/o3de/o3de/pull/8312/commits/5a6f2f99ec06d8f943905176b9b42cf3cc40a13f)

### CODE REVIEW

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:74]: nit: Just naming these xxxCenter, xxxRight, xxxTop would remove the need for comments and make it more clear.

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:94]: is -> if

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:108]: nit: Almost everything from the PassSrg is passed into DirectionalShadowCalculator, but DirectionalShadowCalculator still accesses PassSrg directly for the sampler state. Seems like DirectionalShadowCalculator should either not care about the SRG at all (everything passed in) or be intentionally tightly coupled to the pass srg (almost nothing passed in).

File: Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp:66]: @antonmic Seems like this could be a generic pass system utility.

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp
Comment (invertednormal) [inline Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp:1661]: nit: having a bunch of separate setters exposed which are all just called one after the other doesn't seem like a great pattern. If all of this stuff is expected to be set as part of initialization, I think it'd be cleaner to have a single `Initialize()` function that just takes a descriptor. This would make more sense if those properties were individually set from different places in the code.

**Review by invertednormal — APPROVED** [review]
Looks good - I just have a few nits, but nothing holding up the review.

File: Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp
Comment (antonmic) [inline Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp:66]: You don't need this function, we already have Pass::FindAttachmentBinding

**Review by antonmic — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Source/Shadows/FullscreenShadowPass.cpp:66]: It is, see Pass::FindAttachmentBinding

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:108]: Accessing PassSrg for the sampler state is fine imo, at least for now. I think this is more of a Atom wide issue than a shadows issue. We could probably have basic linear and point samplers in the Scene SRG instead of defining them in every PassSRG, I will raise it in the leads meeting, but definitely beyond the scope of this PR

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.cpp:1661]: Ideally the Pass should get the data from the Feature Processor, not the other way around. But it's fine for now, some other passes/FPs already do this kind of caching thing, should do a general pass over them as some point to change it to passes querying FPs

**Comment by Kadino** [comment]
There don't appear to be new or modified tests as part of this review. Did you have a question for SIG-Testing?

**Comment by Kadino** [comment]
> There don't appear to be new or modified tests as part of this review. Did you have a question for SIG-Testing?

Oh, interesting. It appears that CODEOWNERS automatically added sig-testing-reviewers due to the merge from dev.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli:192]: My memory of applying this technique is that the slope-scaled bias is applied when rendering the shadow map itself (with a known camera pose), and not something applied when sampling the shadow map, mainly since this can be accomplished by changing the rasterizer state and leveraging the hardware. I don't think it's uncommon to rely on a slope bias in the rasterizer state, and subsequently do a view-based bias after also.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli:302]: After seeing this condition a number of times I wonder if it's worth refactoring it to be a helper member function

File: Gems/Atom/Feature/Common/Assets/ShaderResourceGroups/ViewSrg.azsli
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Assets/ShaderResourceGroups/ViewSrg.azsli:15]: Maybe `World -> Clip` is more accurate (I don't know why but "screen" to me has connotations of being scaled by screen extent or strictly screenspace uvs whereas clip space coordinates still possess depth)

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:75]: gather -> gathers

File: Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.h
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/CoreLights/DirectionalLightFeatureProcessor.h:196]: Would be nice to have acceptable ranges for these values in comments as well

**Review by jeremyong-az — APPROVED** [review]

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli:38]: Good catch, thx

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli:44]: Good catch, thx

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli:104]: How would you account for a quad of pixels that is affected by multiple cascades?

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli:156]: Will do

File: Gems/Atom/Feature/Common/Assets/ShaderResourceGroups/ViewSrg.azsli
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/ShaderResourceGroups/ViewSrg.azsli:15]: Good suggestion, will  change

File: Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl
Comment (antonmic) [reply Gems/Atom/Feature/Common/Assets/Shaders/Shadow/FullscreenShadow.azsl:75]: will fix

**Pull request merged**
